url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/event/EventTaskRunnerImple.java

```diff
@@ -0,0 +1,215 @@
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import net.hetimatan.util.log.Log;
+
+public class EventTaskRunnerImple extends EventTaskRunner {
+	private static int sID = 0;
+	public int mId = sID++;
+	private SingleTaskRunner mRunner = new SingleTaskRunner();
+
+	private LinkedList<EventTask> mTasks = new LinkedList<EventTask>();
+	private LinkedList<DefferTask> mDefferTasks = new LinkedList<DefferTask>();
+
+	private Worker mWorker = null;
+
+	public boolean currentThreadIsMine() {
+		return mRunner.currentThreaddIsMime();
+	}
+
+	public EventTaskRunnerImple() {
+	}
+
+	@Override
+	public boolean isAlive() {
+		if(mRunner == null) {return false;}
+		return mRunner.isAlive();
+	}
+
+	public synchronized void pushTask(EventTask task) {
+		start(task);
+	}
+
+	public synchronized void pushTask(EventTask task, int timeout) {
+		mDefferTasks.add(new DefferTask(task, timeout));
+		if(mRunner == null|| mWorker == null) {
+			start(null);
+		}
+		kickWorker();
+	}
+
+	@Override
+	public void releaseTask(EventTask task) {
+		if(task != null) {
+		mDefferTasks.remove(task);
+		}
+	}
+
+	public int updateDeffer() {
+		long curTime = System.currentTimeMillis();
+		long time = -1;
+		long ret = -1;
+		for(int i=0;i<mDefferTasks.size();i++) {
+			DefferTask task = mDefferTasks.get(i);
+			time = task.deffer(curTime);
+			if(time<=0) {
+				pushTask(task.getEventTask());
+				mDefferTasks.remove(i);
+				ret = 0;
+			} else {
+				if(ret==-1) {
+					ret = time;					
+				}
+				else if(ret>time) {
+					ret = time;
+				}
+			}
+		}
+		return (int)ret;
+	}
+
+	public synchronized EventTask popTask() {
+		try {
+		if (mTasks.size() > 0) {
+			return mTasks.remove(0);//removeFirst();
+//			return mTasks.removeFirst();//(0);//removeFirst();
+		} else {
+			return null;
+		}
+		} catch(NoSuchElementException e) {
+			return null;
+		}
+	}
+
+	@Override
+	public void start(EventTask task) {
+		super.start(task);
+		//pushWork(task);
+//		mTasks.addLast(task);
+		if (task != null) {
+			mTasks.add(task);
+		}
+		if (mRunner == null || !mRunner.isAlive()) {
+			mRunner = new SingleTaskRunner();
+			mRunner.startTask(mWorker = new Worker(this));
+		} else if(mWorker != null){
+			//mWorker.kick();
+			kickWorker();
+		}
+	}
+
+	public void dispose() {
+		mRunner.endTask();
+	}
+
+	public void kickWorker() {
+		mWorker.kick();
+	}
+
+	public synchronized void kick() {
+		notify();
+	}
+	public synchronized void waitPlus(int time) throws InterruptedException, IOException {
+		if(time<0) {
+			//Thread.sleep(10000);
+			wait(10000);
+		} else {
+			if(time > 0) {
+				wait(time);
+			}
+			//Thread.sleep(time);			
+		}
+	}
+
+	public static class Worker implements Runnable {
+
+		WeakReference<EventTaskRunnerImple> mRunner = null;
+
+		public Worker(EventTaskRunnerImple runner) {
+			mRunner = new WeakReference<EventTaskRunnerImple>(runner);
+		}
+
+		public void kick() {
+			EventTaskRunnerImple runner = mRunner.get();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *runner*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).